src/action-providers.ts

```diff
@@ -21,18 +21,87 @@ class RefactoringActionProvider implements vscode.CodeActionProvider {
     const ast = t.parse(document.getText());
     const selection = createSelectionFromVSCode(range);
 
-    return this.refactorings
-      .filter(refactoring => this.canPerform(refactoring, ast, selection))
-      .map(refactoring => this.buildCodeActionFor(refactoring));
+    const applicableRefactorings: RefactoringWithActionProvider[] = [];
+
+    const onCanPeform = (
+      path: t.NodePath<any>,
+      refactoring: RefactoringWithActionProvider
+    ) => {
+      if (isLegacyActionProvider(refactoring.actionProvider)) {
```

Comment:
As mentioned in the `canPerformLegacy` function: I'd split the legacy and non-legacy refactorings first, and then map them to the appropriate function. That way, we don't have to duplicate the checks everywhere (e.g. here and in `canPerform()`).

What do you think?